Add performance warning for too many case properties

gherceg opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
https://www.loom.com/share/62f8465ff8fc4275934096ca2717daa4

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-17714

Review by commit 🐟 

We want to provide a warning to users once they hit a certain number of case properties on a case type. This is set to 250 in this PR but is something we can change per environment if needed.

The idea is that users might add case properties without thinking twice about it, so this provides a warning for them to re-evaluate their usage of case properties, link to some docs, and possibly alter their path.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
The blast radius is pretty low given this is an isolated alert that if it breaks, shouldn't cause issues on the page itself. I've tested this on staging quite a bit and it behaves as expected.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Add SystemLimit key for case property limit

gherceg committed:
Implement warning on case management page

gherceg committed:
Implement warning on data dictionary

gherceg committed:
Fix formatting errors

gherceg committed:
Use data dictionary as source of truth

Data dict models get updated for all plans

gherceg committed:
Update warning content

gherceg committed:
Lint JS

I left one in place because adding a trailing comma to the string didn't
make sense to me.

jingcheng16 commented:
Haven't reviewed the PR yet, just looked at the video.
Do you know what will happen after user delete a case property, if there exists a mapping between a question and a case property in the case management settings page? Will that mapping be deleted too?
My concern is, if that mapping is not deleted, then once the app containing the mapping is saved again, we will sync the case property defined in the case management settings page to the data dictionary again.
My concern is out of scope of this PR but just curious.

millerdev commented on corehq/apps/app_manager/views/forms.py:
```diff
@@ -809,6 +812,12 @@ def get_form_view_context(
         'module_loads_registry_case': module_loads_registry_case(module),
+        'case_property_warning': {
+            'count': case_property_count,
+            'limit': _get_case_property_limit(domain),
+            'show': case_property_count > _get_case_property_limit(domain),
```
I assume calling `_get_case_property_limit()` twice is not likely to cause a performance issue, but seems kinda wasteful since it's a DB hit (or two). Maybe worth storing the result in a variable to avoid the second call?

millerdev commented on corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:
```diff
@@ -606,6 +613,33 @@ var dataDictionaryModel = function (dataUrl, casePropertyUrl, typeChoices, fhirR
+        if (caseType.propertyCount > self.casePropertyLimit) {
+            let content = gettext(
+                "The '" + caseType.name + "' case has " + caseType.propertyCount + " properties. " +
+                "We recommend keeping case properties below " + self.casePropertyLimit + ", " +
```
Translating a string containing pre-substituted variable values will not work, right?

gherceg replied:
[512de98](https://github.com/dimagi/commcare-hq/pull/36588/commits/512de98a6a4c9cf34ed503afc20941d5dbd81937)

millerdev commented on corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:
```diff
@@ -606,6 +613,33 @@ var dataDictionaryModel = function (dataUrl, casePropertyUrl, typeChoices, fhirR
+        let icon = toggle.querySelector('i');
+        if (icon.classList.contains('fa-chevron-down')) {
+            icon.classList.remove('fa-chevron-down');
+            icon.classList.add('fa-chevron-up');
```
Does changing the class on the icon trigger expand/collapse? What prevents the chevron direction from getting out of sync with expanded/collapsed state?

Should `toggleCasePropertyWarning` change the value of `showCasePropertyWarning`, or does it do that through some magical convolution that I'm missing?

If there is an issue here, there also may be an issue [over here](https://github.com/dimagi/commcare-hq/pull/36588/commits/1e1b2e0e08a26f8535d831c7edd4a23e92f1713f#diff-e511cabdd447893388e1a65c371c787f2f7c94e76697607c307131e0cdb587a5).

jingcheng16 replied:
+1

gherceg replied:
I ended up using the `collapsed` class on the toggle, which I confirmed gets added and removed automatically. The hack is that I needed to explicitly add it so that it matches that starting state of the element.

[f845770](https://github.com/dimagi/commcare-hq/pull/36588/commits/f8457704c85ce848d24b3db2e335ac067773edad)

millerdev commented on corehq/apps/app_manager/templates/app_manager/form_view.html:
```diff
@@ -181,8 +181,8 @@
+                    The '{{ type }}' case has a total of {{ count }} custom properties. We
+                    recommend at most {{ limit }} custom properties per case type, otherwise
                     you may run into performance issues at the time of data
                     collection and analysis.
```
Should this be kept in sync with this other similar message?

https://github.com/dimagi/commcare-hq/blob/0a27457d9c285f02e64c5cb283ebd000a857fa23/corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js#L620-L622

jingcheng16 commented on corehq/project_limits/const.py:
```diff
@@ -1,3 +1,6 @@
 DEVICE_LIMIT_PER_USER_KEY = "device_limit_per_user"
+
+# SystemLimit defaults
+DEFAULT_CASE_PROPERTIES_PER_CASE = 250
```
"per case type" ?

gherceg replied:
[3019b1a](https://github.com/dimagi/commcare-hq/pull/36588/commits/3019b1aaf90388e60b3a8b94f190a627d70229b6)

jingcheng16 commented on corehq/apps/app_manager/static/app_manager/js/forms/form_view.js:
```diff
@@ -170,6 +170,15 @@ $(function () {
     // Case Management > Data dictionary descriptions for case properties
     $('.property-description').popover();
 
+    $('#performance-warning-toggle').on('click', function () {
```
This feels a little bit weird to me that we toggles the icon class based on its current class when the button is clicked not the actual collapse state.
What do you think of something like:
```
$('#warningContent').on('show.bs.collapse', function () {
    $('#performance-warning-toggle i')
        .removeClass('fa-chevron-down')
        .addClass('fa-chevron-up');
});

$('#warningContent').on('hide.bs.collapse', function () {
    $('#performance-warning-toggle i')
        .removeClass('fa-chevron-up')
        .addClass('fa-chevron-down');
});
```

gherceg replied:
I ended up using the `collapsed` class on the toggle, which I confirmed gets added and removed automatically. The hack is that I needed to explicitly add it so that it matches that starting state of the element.

[f845770](https://github.com/dimagi/commcare-hq/pull/36588/commits/f8457704c85ce848d24b3db2e335ac067773edad)

gherceg replied:
Just kidding, ended up going with your approach! [08d33a1](https://github.com/dimagi/commcare-hq/pull/36588/commits/08d33a11cdb8680910a90b08344a5d77cdbf95e4)

jingcheng16 commented on corehq/apps/app_manager/templates/app_manager/form_view.html:
```diff
@@ -166,6 +165,38 @@
         <div class="tab-pane" id="case-configuration">
+          {% if case_property_warning.show %}
+            <div class="alert alert-warning">
+              <button
```
Wondering if only the small button (#performance-warning-toggle) is clickable? I feel make the entire header line clickable is more user-friendly.

gherceg replied:
Yeah only the button is clickable. I'd be curious to hear your feedback after playing with it on staging.

jingcheng16 commented on corehq/apps/app_manager/templates/app_manager/form_view.html:
```diff
@@ -166,6 +165,38 @@
+              <button
+                id="performance-warning-toggle"
+                type="button"
+                class="close"
```
This doesn't sit right with me. Are you giving it class = "close" just to have it positioned in the right top corner? But this is not actually a close action. I feel this is hack but if we already do it in other places I won't be too strict on it.

gherceg replied:
Yeah it positions it in the top right corner. I didn't know of another straightforward way to do it, but agree it is a bit of a hack.

jingcheng16 commented on corehq/apps/data_dictionary/templates/data_dictionary/base.html:
```diff
@@ -153,6 +154,28 @@ <h4 class="modal-title">{% trans "Create a new Case Type" %}</h4>
           <h3 data-bind="text: $root.activeCaseType()" style="display: inline-block;"></h3>
           <span data-bind="visible: $root.isActiveCaseTypeDeprecated()" class="deprecate-case-type hidden label label-warning" style="display: inline-block;">{% trans "deprecated" %}</span>
         </div>
+        <div class="alert alert-warning" data-bind="visible: showCasePropertyWarning()">
```
This looks very similar to what we have in case management page, I didn't do a detailed comparison. But if they're the same, does it make sense to make it a partial?

gherceg replied:
Ended up making the partial work! [8ee6103](https://github.com/dimagi/commcare-hq/pull/36588/commits/8ee6103996bd0483f76a69ba9922bca8006bd489)

jingcheng16 reviewed (changes requested):
One additional comment regarding your commit 5acfe33:
Unfortunately DD cannot be treated as source of truth yet.  ><

gherceg committed:
Make it clear that limit is per case type

As opposed to "case" which has another meaning in HQ

gherceg committed:
Use underscore template for interpolated translation

gherceg committed:
Determine chevron icon based on collapsed class

gherceg commented on corehq/project_limits/const.py:
```diff
@@ -1,2 +1,3 @@
 # SystemLimit keys
+CASE_PROPERTIES_PER_CASE_KEY = "case_properties_per_case"
```
[3019b1a](https://github.com/dimagi/commcare-hq/pull/36588/commits/3019b1aaf90388e60b3a8b94f190a627d70229b6)

gherceg committed:
Move html and javascript into reusable partial

gherceg committed:
Use bootstrap collapse events to toggle icon

gherceg committed:
Slight rewording of warning

gherceg committed:
JS linting

gherceg commented on corehq/apps/app_manager/static/app_manager/js/forms/form_view.js:
```diff
@@ -1,7 +1,7 @@
 import $ from "jquery";
 import ko from "knockout";
-import _ from "underscore";
+import _, { initial } from "underscore";
```
This was an accident, and is removed in a later commit.

millerdev reviewed (approved):
LGTM. Not sure if @jingcheng16 will want to review yet as well?

I really like how https://github.com/dimagi/commcare-hq/pull/36588/commits/8ee6103996bd0483f76a69ba9922bca8006bd489 eliminated duplicate code.

gherceg committed:
Remove unnecessary collapsed class

This was added for a previous soluton in commit f845770 that I should
have removed in 08d33a1.

gherceg commented on corehq/apps/data_dictionary/templates/data_dictionary/base.html:
```diff
@@ -155,19 +155,19 @@ <h3 data-bind="text: $root.activeCaseType()" style="display: inline-block;"></h3
-          <i class="fa fa-exclamation-triangle"></i>
-          {% trans "Performance warning" %}
-          <div id="warningContent" class="collapse">
+          <div id="performance-warning-content" class="collapse">
```
[9efed9d](https://github.com/dimagi/commcare-hq/pull/36588/commits/9efed9da09c8b68f159675077216cfaff08800be)

jingcheng16 reviewed (approved):
Looks good. As Daniel mentioned, glad to see you made 8ee6103996bd0483f76a69ba9922bca8006bd489! 
Agree the worst case is we didn't show the warning for the case we should show.
